Add coverage on pkg/archive (related to #11603)

vdemeester opened the pull request:
Add tests on:
- changes.go
- archive.go
- wrap.go

Should fix #11603 as the coverage is now 81.2% on the `pkg/archive`
package. There is still room for improvement though :).

Signed-off-by: Vincent Demeester vincent@sbr.pm

vdemeester commented:
ah one test is failing on jenkins.. hum :(

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	}
+	defer os.RemoveAll("/tmp/dest")
+	err = UntarPath("", "/tmp/dest")
+	if err == nil {
```
Its not clear to me what you're testing here. The Fatal() message and name of the test would seem to indicate that you're testing for an invalid src but it could have also failed due to you pointing to a destination file instead of a destination dir.  I would recommend that you split this into two.  One to validate that it generates an error on a bad "src" and one for testing a bad "dest" - but in both cases make sure the other param passed in is valid.

vdemeester replied:
You're right, will do.

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+}
+
+func TestUntarPath(t *testing.T) {
+	cmd := exec.Command("/bin/sh", "-c", "touch /tmp/src && tar cfT /tmp/src.tar /tmp/src")
```
I would recommend that you create a new/temp dir for stuff like this otherwise you may end up with a conflict with pre-existing files on the system.

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	defer os.RemoveAll("/tmp/dest")
+	err = UntarPath("/tmp/src.tar", "/tmp/dest")
+	if err != nil {
+		t.Fatalf("UntarPath shouldn't throw an error, %s.", err)
```
I wonder why this shouldn't fail since dest isn't a dir - its file. Even if the tar is empty (like you created) I kind of expected things to blow up.

vdemeester replied:
I should have write different test but it's related to #10040. There is a comment that state :

```go
		// If path exits we almost always just want to remove and replace it
		// The only exception is when it is a directory *and* the file from
		// the layer is also a directory. Then we want to merge them (i.e.
		// just apply the metadata from the layer).
```

vdemeester replied:
But it should be more clear in the test, so I'm gonna write one with a file to mark it.

vdemeester replied:
Ah that's trickier than that ! Thanks for you comment, it didn't work exactly as I though it was..

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+		t.Fatalf("UntarPath shouldn't throw an error, %s.", err)
+	}
+	_, err = os.Stat("/tmp/src")
+	if err != nil {
```
Not sure what you're testing here.  You created /tmp/src via the touch, you don't delete it (except via a defer for later processing), so why would this test ever fail? And what did this have to do with the untar?  Did you mean to test for /tmp/dest/src instead? If so, then you'll need to fix your tar because the tar file is empty based on how you created it.

vdemeester replied:
That is right, messed up my test there :'(

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+func TestCopyWithTarInvalidSrc(t *testing.T) {
+	err := CopyWithTar("", "/tmp/dest")
+	if err == nil {
+		t.Fatalf("archiver.CopyWithTar with invalid src path should throw an error.")
```
Like above, let's make sure we only pass in one bad thing at a time to make sure we get the correct error.  Let's split this into two - one for bad tar file and one for bad dest dir.

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	if err != nil {
+		t.Fatalf("archiver.CopyWithTar shouldn't throw an error, %s.", err)
+	}
+	_, err = os.Stat(src)
```
I think 'src' should be path.Join(dest, "src") or something like that.

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	defer os.RemoveAll(folder)
+	dest := path.Join(folder, "dest")
+	srcFolder := path.Join(folder, "src")
+	src := path.Join(folder, "src/src")
```
You may need to use filepath.Join() and split "src/src" into "src", "src" so that it uses the appropriate / vs \\ for when its run on Linux vs Windows.

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	}
+	defer os.RemoveAll(folder)
+	dest := path.Join(folder, "dest")
+	src := path.Join(folder, "src/folder")
```
See above comments about / vs \\

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	if err != nil {
+		t.Fatalf("archiver.CopyWithTar shouldn't throw an error, %s.", err)
+	}
+	_, err = os.Stat(src)
```
check for 'dest'/... instead of 'src'

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+
+func TestCopyFileWithTarInvalidSrc(t *testing.T) {
+	err := CopyWithTar("/bad", "/tmp/dest")
+	if err == nil {
```
see comments above about splitting this test into 2 so we're clear on what we're testing.

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = CopyFileWithTar(src, dest)
```
I'm not sure this test is any different from the previous one because both are passing in a folder for the src. Is one of them wrong and should be passing in a file instead?

vdemeester replied:
This one should fail (test just below). If ``CopyFileWithTar`` is called with a folder, it will throw an error. (there was a ``os.Stat`` after that was there for no reason..). I think I was a little bit tired when I wrote this one :D

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	if err == nil {
+		t.Fatalf("CopyFileWithTar should throw an error with a folder.")
+	}
+	_, err = os.Stat(src)
```
check dest instead of src

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	if err != nil {
+		t.Fatalf("archiver.CopyFileWithTar shouldn't throw an error, %s.", err)
+	}
+	_, err = os.Stat(src)
```
checking src by mistake

duglin commented on pkg/archive/archive_test.go:
```diff
@@ -207,6 +207,164 @@ func TestCmdStreamGood(t *testing.T) {
+	defer os.RemoveAll(folder)
+	dest := path.Join(folder, "dest")
+	srcFolder := path.Join(folder, "src")
+	src := path.Join(folder, "src/src")
```
use filepath instead of path and use srcFolder instead of folder, then you can just pass in "src" instead of "src/src".

duglin commented on pkg/archive/changes_test.go:
```diff
@@ -91,17 +92,130 @@ func createSampleDir(t *testing.T, root string) {
+	}
+	defer os.RemoveAll(layer)
+	createSampleDir(t, layer)
+	changes, err := Changes([]string{layer}, rwLayer)
```
I'm not familiar with this Changes() code so ignore me if I'm totally wrong, but don't we need to do something to associate these two dirs/layers (some kind of aufs magic?) or is just passing in the slice of parent layers enough?

vdemeester replied:
Well from my understanding (hoping I'm not wrong either), the code here works with any "folder", backed by aufs or not. Because the filesystem backend is not always aufs (could be btrfs & co) so :).
The only "tricky stuff" is the deletion of files where aufs creates a ``.wh.file1-2`` to tell aufs it has been deleted.

vdemeester committed:
Add coverage on pkg/archive

Add tests on:
- changes.go
- archive.go
- wrap.go

Should fix #11603 as the coverage is now 81.2% on the ``pkg/archive``
package. There is still room for improvement though :).

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
I've updated my commit according to @duglin comments (thanks a lot.. there was a lot of mess :-/). I added few FIXME to test some more. I will update it later (in a few hour).

calavera commented:
tests LGTM :muscle:

LK4D4 commented:
@vdemeester Thank you. This is very important work!
LGTM
